Library/Homebrew/software_spec.rb

```diff
@@ -173,6 +173,7 @@ def deprecated_option(hash)
   end
 
   def depends_on(spec)
+    remove_gnu_compilers(spec)
```

Comment:
Interesting hack, and not crazy. My preference would be to add `fails_with gcc: "5"` to the formula, and allow the compiler selector to select the correct compiler. I believe `brew test-bot` is responsible for installing the necessary compiler.

Reply:
Yeh, I think I agree here.

Reply:
> My preference would be to add fails_with gcc: "5" to the formula, and allow the compiler selector to select the correct compiler

Can you be a little bit more explicit about this, I am not sure to understand what you mean.

- On Mac a plain `depends_on "gcc"` was added in 84 formulae (often related to the need to have gfortran with gcc 10), and I do not want to need to add `fails_with gcc: "5` everywhere in a `on_linux` block, I just want brew to install and use gcc 10 in that case.
- On linux some formulae do not build with gcc5 (they do build with llvm on Mac though so nobody noticed expect the linux people): in that case the plan is to also use `depends_on "gcc"` if possible, but just for linux.

This does not change anything to the fact that for all the other formulae where `depends_on "gcc"` was NOT added to the formula, `/usr/bin/gcc-5` will be used by default.

Reply:
> * On Mac a plain `depends_on "gcc"` was added in 84 formulae (often related to the need to have gfortran with gcc 10), and I do not want to need to add `fails_with gcc: "5` everywhere in a `on_linux` block, I just want brew to install and use gcc 10 in that case.
> * On linux some formulae do not build with gcc5 (they do build with llvm on Mac though so nobody noticed expect the linux people): in that case the plan is to also use `depends_on "gcc"` if possible, but just for linux.

What's the behaviours before this PR/without this change and no `fails_with`?

> This does not change anything to the fact that for all the other formulae where `depends_on "gcc"` was NOT added to the formula, `/usr/bin/gcc-5` will be used by default.

So the issue is: GCC 5 being prioritised even when it's depended on directly?

Reply:
> What's the behaviours before this PR/without this change and no fails_with?

ENV.cc and ENV.cxx will not point to the right gcc version.
For example you can add `depends on gcc@9` to a formula:
- if no brewed gcc is installed; ENV.cc will point to system gcc-5 on linux, which I think is wrong is that case, as I requested gcc@9 in the formula
- if brewed gcc is installed, ENV.cc points to that gcc (gcc-5 now or gcc-10 in the future), which is also wrong, as I request gcc@0 in the formula

You need to: fix the "preferred gcc" part first to get it right, and add all the necessary fails_with to finally get the expected result.

> So the issue is: GCC 5 being prioritised even when it's depended on directly?

Err I am unsure I understand what you mean by this. I think the sentence above misses a negation somewhere.

Reply:
> On Mac a plain depends_on "gcc" was added in 84 formulae (often related to the need to have gfortran with gcc 10), and I do not want to need to add fails_with gcc: "5 everywhere in a on_linux block, I just want brew to install and use gcc 10 in that case.

Change `depends_on "gcc"` to
```ruby
on_macos do
  depends_on "gcc"
end
```

> On linux some formulae do not build with gcc5 (they do build with llvm on Mac though so nobody noticed expect the linux people): in that case the plan is to also use depends_on "gcc" if possible, but just for linux.

Add `fails_with gcc: "5"`

That's my suggestion. Happy to discuss further.

Reply:
I also agree with @sjackman that I think it's worth decoupling the potential syntactical improvements from the wider goals of this PR.

Reply:
> Change depends_on "gcc" to `on_macos do depends_on "gcc" end`

It's a little bit more complex than that. There are extra flags added here and there for gcc 10 in the Mac formulae (for example -fallow-argument-mismatch, and maybe other stuff). So there might be more stuff to handle in these formulae.
I would really like to keep the gcc versions chosen for a formula as close as possible on Mac and linux.

> Add fails_with gcc: "5"

Without this PR, you need to add the whole list of fails_with for each version of gcc you want to exclude.

> I also agree with @sjackman that I think it's worth decoupling the potential syntactical improvements from the wider goals of this PR.

The only thing I can do is split this PR in two: on for the preferred gcc, one for the fails_with hack. Is this what you mean?

Reply:
> Without this PR, you need to add the whole list of fails_with for each version of gcc you want to exclude.

This seems right/expected. What should change is the preference for the GCC version that's selected but this shouldn't require looking at the dependences at all.



> The only thing I can do is split this PR in two: on for the preferred gcc, one for the fails_with hack. Is this what you mean?

Yes, I think so.

Reply:
>> Add fails_with gcc: "5"

> Without this PR, you need to add the whole list of fails_with for each version of gcc you want to exclude.

Homebrew on Linux CI has only two compilers available: `/usr/bin/gcc-5` and `/home/linuxbrew/.linuxbrew/bin/gcc-10`. No formulae in `Homebrew/core` depend on `gcc@*`, so other versions of GCC will never be installed by CI. `fails_with gcc: "5"` is sufficient to tell `CompilerSelector` to use `gcc-10` rather than `gcc-5`.

Reply:
>> Change depends_on "gcc" to on_macos do depends_on "gcc" end

> It's a little bit more complex than that. There are extra flags added here and there for gcc 10 in the Mac formulae (for example -fallow-argument-mismatch, and maybe other stuff). So there might be more stuff to handle in these formulae.
I would really like to keep the gcc versions chosen for a formula as close as possible on Mac and linux.

In that case add `fails_with gcc: "5"`.

If the formula builds with GCC 10 but neither GCC 5 nor Clang on macOS:

```ruby
depends_on "gcc"
fails_with :clang
fails_with gcc: "5"
```

If the formula builds with both GCC 5 and GCC 10 but not Clang on macOS:

```ruby
on_macos do
  depends_on "gcc"
end
fails_with :clang
```

If the formula builds with both Clang on macOS and GCC 10 but not GCC 5:

```ruby
on_linux do
  depends_on "gcc"
end
fails_with gcc: "5"
```